domain/upgrade/state/state.go

```diff
@@ -0,0 +1,304 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/collections/set"
+	"github.com/juju/collections/transform"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+	"github.com/juju/version/v2"
+
+	"github.com/juju/juju/domain"
+)
+
+// State is used to access the database.
+type State struct {
+	*domain.StateBase
+}
+
+// noUpgradeErr is the error returned if the user attempts to
+// perform operations when no upgrade is in progress
+var noUpgradeErr = errors.Errorf("no current upgrade")
+
+// NewState creates a state to access the database.
+func NewState(factory domain.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+func (st *State) getCurrentUpgrade(ctx context.Context, tx *sqlair.TX) (info, []infoControllerNode, error) {
```

Comment:
This could be done in one query with a `MAX`, `GROUP BY` and `JOIN` if supported by SQLair.

Reply:
This jhas been made irrelevant by other changes